### DISCUSSION TITLE
Zettel change history

### DISCUSSION THREAD

**Discussion by applejag** [srid/neuron#546] [Ideas]
Since it's common to version control neuron via Git, you do get some "free" version history of all the zettels.

I see two use cases for this.

1. Displaying a change history at the bottom of a zettel. Maybe collapsed by default, but still available, where it's more or less just taking the git commit history for the file. Things like when it was changed and by whom. Maybe even linking to the Git commit over at GitHub if you have that integration set up.

2. Using Git for this is nice because Git also keeps tracks of file renames if you supply `git log` with the `--follow` argument. Using this, you could potentionally generate page redirects for renamed zettels. Just a basic page showing "This page was moved to: the-new-url.html".

   There's the edge case as well where you rename file A.md to B.md, and then from B.md to C.md. Then you rename some other file, D.md, to B.md. Given the history of the files (which Git keeps track of, still just using the `--follow` argument), you could add a small notice at the top or bottom of this new B.md (prev D.md) file saying "This page was moved and replaced. Looking for C.html?"

   This would allow better futureproofing by not discouraging renames of zettels.

---

Talking implementation then this would obviously be a plugin of sort. I do not know what other SCM people are using for neuron but there are probably some other variants out there so making it extensible with for example Plastic SCM or mercurial instead would probably be a good idea.

**Comment by applejag** [comment]
I would very much like to try implementing this, but I've no idea where to start. @srid do you have any guides on contributing? I've seen the [CONTRIBUTING.md](https://github.com/srid/neuron/blob/master/CONTRIBUTING.md), but it only goes through how to setup the development environment.

Do you have any tips on where to begin the journey of adding plugins, or just basically "read the source code until you get it"?

**Comment by srid** [comment]
@jilleJr I would be happy to help. [Matrix chat](https://app.element.io/#/room/#neuron:matrix.org) is a good place to get fast dev help. 

How familiar are you with Haskell?

The plugin API is very much still a WIP, but you can start from any of the plugins [here](https://github.com/srid/neuron/tree/master/neuron/src/lib/Neuron/Plugin/Plugins). After that, it is very much "read the source code until you get it" (with some help from chat).

@jared-w had some ideas on creating a `mainWithPlugin` to allow using neuron as a library, so you can write your own plugins in your own repo. But that will require some refactoring, which @jared-w expressed some interest on (and I can absolutely help here).

**Reply by applejag** [reply]
I'm not that familiar with Haskell, but I'm familiar with F#. The idea here is that hopefully that knowledge is transferrable, but I've wanted to peak at haskell for a while now. I've zero knowledge about the haskell eco-system or the standard libraries, so that's what will have to take some time to get up to speed with first.

If I attack this idea, there will be no quick development cycle. Slow and steady.

**Comment by srid** [comment]
As an aside, I'm experimenting with an entirely new model of data transformation (and plugins) based on Unix pipelines ([see here](https://github.com/srid/g/blob/53a166370c9620f94d60d7b6b34d1725e4bfc330/src/G/Pipeline.hs#L54-L109)) -- [g](https://github.com/srid/g). `g` might become a core of neuron; but right now, it is a playground to explore the problem space without worrying about backwards compat.

**Reply by bew** [reply]
You got me interested, what kind of things do you have in mind?

**Reply by srid** [reply]
@bew I'll write about it in detail later, but for now - you can get it running (see README; I just pushed), check the docs it serves, and look around the source code (and even try editing it; the dev server will reload instantly).
